src/FSharp.Data.GraphQL.Server.AppInfrastructure/GraphQLWebsocketMiddleware.fs

```diff
@@ -0,0 +1,375 @@
+            | Direct (data, _) ->
+                do! data
+                    |> safe_SendQueryOutput id
+        }
+
+    let getStrAddendumOfOptionalPayload optionalPayload =
+        optionalPayload
+        |> Option.map (fun payloadStr -> sprintf " with payload: %A" payloadStr)
+        |> Option.defaultWith (fun () -> "")
+
+    let logMsgReceivedWithOptionalPayload optionalPayload (msgAsStr : string) =
+      if logger.IsEnabled(LogLevel.Trace) then
+        logger.LogTrace ("{message}{messageaddendum}", msgAsStr, (optionalPayload |> getStrAddendumOfOptionalPayload))
+
+    let logMsgWithIdReceived (id : string) (msgAsStr : string) =
+      if logger.IsEnabled(LogLevel.Trace) then
+        logger.LogTrace("{message} (id: {messageid})", msgAsStr, id)
+
+    // <--------------
+    // <-- Helpers --|
+    // <--------------
+
+    // ------->
+    // Main -->
+    // ------->
+    task {
+      try
+        while not cancellationToken.IsCancellationRequested && socket |> isSocketOpen do
+            let! receivedMessage = safe_Receive()
+            match receivedMessage with
+            | None ->
+              if logger.IsEnabled(LogLevel.Trace) then
+                logger.LogTrace(sprintf "Websocket socket received empty message! (socket state = %A)" socket.State)
+            | Some msg ->
+                match msg with
+                | Failure failureMsgs ->
+                    "InvalidMessage" |> logMsgReceivedWithOptionalPayload None
+                    match failureMsgs |> List.head with
+                    | InvalidMessage (code, explanation) ->
+                      do! socket.CloseAsync(enum code, explanation, new CancellationToken())
+                | Success (ConnectionInit p, _) ->
+                    "ConnectionInit" |> logMsgReceivedWithOptionalPayload p
+                    do! socket.CloseAsync(
+                      enum CustomWebSocketStatus.tooManyInitializationRequests,
+                      "too many initialization requests",
+                      new CancellationToken())
+                | Success (ClientPing p, _) ->
+                    "ClientPing" |> logMsgReceivedWithOptionalPayload p
+                    match pingHandler with
+                    | Some func ->
+                      let! customP = p |> func serviceProvider
+                      do! ServerPong customP |> safe_Send
+                    | None ->
+                      do! ServerPong p |> safe_Send
+                | Success (ClientPong p, _) ->
+                    "ClientPong" |> logMsgReceivedWithOptionalPayload p
+                | Success (Subscribe (id, query), _) ->
+                    "Subscribe" |> logMsgWithIdReceived id
+                    if subscriptions |> GraphQLSubscriptionsManagement.isIdTaken id then
+                      do! socket.CloseAsync(
+                        enum CustomWebSocketStatus.subscriberAlreadyExists,
+                        sprintf "Subscriber for %s already exists" id,
```

Comment:
logger

Reply:
Done (now in file `src/FSharp.Data.GraphQL.Server.AspNetCore/GraphQLWebsocketMiddleware.fs`).